Literal: General

Hywan opened the pull request:
Address https://github.com/tagua-vm/parser/issues/5.
### Specification

https://github.com/php/php-langspec/blob/master/spec/19-grammar.md#general-3
### Progression
- [x] Create the `Literal` enum in the AST (because all literal rules are returning a different type, we cannot work with that),
- [x] The `null` rule must computes a `Literal`,
- [x] The `boolean` rule must computes a `Literal`,
- [x] The `binary` rule must computes a `Literal`,
- [x] The `octal` rule must computes a `Literal`,
- [x] The `decimal` rule must computes a `Literal`,
- [x] The `hexadecimal` rule must computes a `Literal`,
- [x] The `exponential` rule must computes a `Literal`,
- [x] The `string` rule must computes a `Literal`,
- [x] The `identifier` rule must computes a `Literal`,
- [x] Create the `integer` rule,
- [x] Create the `floating` rule,
- [x] Create the `literal` rule.

Hywan committed:
AST: Add the `Literal` enum.

Hywan committed:
Literal: The `null` rule computes a `Literal`.

Hywan committed:
Literal: The `boolean` rule computes a `Literal`.

Hywan commented on source/ast.rs:
```diff
@@ -55,3 +55,20 @@ pub struct Addition {
     pub b: Term
 }
+
+/// A literal.
```
I am sure we can do better with this documentation 😝.

jubianchi replied:
"A literal represents a fixed value in the source code. For example, a string literal would be `'foobar'`."

jubianchi commented on source/rules/literals.rs:
```diff
@@ -42,25 +42,26 @@ use nom::{
-    pub null< Option<()> >,
+    pub null<Literal>,
     map_res!(
         tag!("null"),
```
I'm afraid this will be case sensitive. Should we use `alt!(tag!("null") | tag!("NULL"))` ?

Hywan replied:
Even `NuLl` is valid :wink:.
Good catch. I will address as a new issue to not pollute this PR.

Hywan replied:
A new item has been added to list in https://github.com/tagua-vm/parser/issues/5.

jubianchi commented on source/rules/literals.rs:
```diff
@@ -42,25 +42,26 @@ use nom::{
-    pub boolean<bool>,
+    pub boolean<Literal>,
     map_res!(
         alt!(tag!("true") | tag!("false")),
```
same thing here.

Hywan replied:
See https://github.com/tagua-vm/parser/pull/24#discussion_r70397778.

Hywan committed:
Literal: The `binary` rule computes a `Literal`.

Hywan committed:
Literal: The `octal` rule computes a `Literal`.

Hywan committed:
Quality: Fix CS.

Hywan committed:
Literal: The `decimal` rule computes a `Literal`.

Hywan committed:
AST: Add explicit lifetimes to each node.

Hywan committed:
Test: Update to include `Literal::Integer`.

Hywan committed:
Literal: The `hexadecimal` rule computes a `Literal`.

Hywan committed:
Literal: The `exponential` rule computes a `Literal`.

Hywan committed:
Literal: The `string` rule computes a `Literal`.

Hywan committed:
Literal: The `identifier` rule computes a `Literal`.

And `Literal::Identifier` is a `Vec<u8>`, thus removing the explicit
lifetime in the AST.

Hywan committed:
Literal: Add the `integer` rule.

Hywan committed:
Literal: Add the `literal` rule.

Hywan committed:
AST: Move `Literal::Float` to `Literal::Real`.

This is important to have the same naming everywhere.

Hywan committed:
Documentation: Complete `Literal` API.

Hywan commented:
@jubianchi: This PR is ready for a review. I guess the `null` & co. case-insensitivity will be solved in #25 and we may remove them from `literal.rs`, but for now, I think this is good to keep them here.

Hywan commented:
@jubianchi: I am not sure if `identifier` must be a `Literal::Identifier`. I think this is a “literal” (I mean, its place is in `literal.rs`) but this is not an alternative of the `literal` rule. What do you think?

jubianchi commented:
@Hywan to me, an identifier is not a literal at all...it's an identifier, a container of something. I'm ok to keep it in the `literal.rs` but it should probably not be part of the `Literal` enum.

The only thing I'm wondering is how much work it will require if we exclude the `Identifier` from the enum. For expressions, any term might be a literal or an identifier. If we can't generalize both it will be hard to describe the parsers I think.

Hywan committed:
Identifier: Extract it from `literals.rs`.

Hywan commented:
@jubianchi Actually, an identifier is part of [the Names Section of the Lexical Structure chapter](https://github.com/php/php-langspec/blob/master/spec/09-lexical-structure.md#names). So this is possible we might rename `identifier.rs` to `names.rs` in a near future. But this is fine for now.
Initially the `identifier` rule computed a `&[u8]`, so absolutely zero-copy, but with the `Literal::Identifier` I had to copy it to a `Vec<u8>`. By extracting it, it was easier for me to understand the issue with the lifetime and now `identifier` computes back a `&[u8]`. So zero-copy again, this is a good news.

I guess we are good now.

Hywan committed:
Test: Simplify a test case.

Hywan committed:
Quality: Fix CS.

Hywan commented:
@jubianchi Ready for anoher review!
